File: src/models/entities.py
"""
Domain entity models
"""

from __future__ import annotations

from datetime import datetime  # noqa: TC003
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class MessageType(str, Enum):
    """Message type enumeration"""

    TEXT = "text"
    MULTIMODAL = "multimodal"
    IMAGE = "image"
    AUDIO = "audio"


class MessageRole(str, Enum):
    """Message role enumeration"""

    USER = "user"
    ASSISTANT = "assistant"


class LastMessageInfo(BaseModel):
    """Last message information for conversation list"""

    model_config = ConfigDict(from_attributes=True, use_enum_values=True)

    content: str | None = None
    role: MessageRole
    created_at: datetime


class InfluencerStatus(str, Enum):
    """Influencer status enumeration"""

    ACTIVE = "active"
    COMING_SOON = "coming_soon"
    DISCONTINUED = "discontinued"


class AIInfluencer(BaseModel):
    """AI Influencer entity"""

    model_config = ConfigDict(from_attributes=True)

    id: str  # Changed from UUID to str to support IC Principal IDs
    name: str
    display_name: str
    avatar_url: str | None = None
    description: str | None = None
    category: str | None = None
    system_instructions: str
    personality_traits: dict[str, object] = Field(default_factory=dict)
    initial_greeting: str | None = None
    suggested_messages: list[str] = Field(default_factory=list)
    is_active: InfluencerStatus = InfluencerStatus.ACTIVE
    is_nsfw: bool = Field(default=False, description="Whether this influencer handles NSFW content")
    parent_principal_id: str | None = None
    source: str | None = None
    created_at: datetime
    updated_at: datetime
    metadata: dict[str, object] = Field(default_factory=dict)

    conversation_count: int | None = None


class Conversation(BaseModel):
    """Conversation entity"""

    model_config = ConfigDict(from_attributes=True)

    id: str  # Changed from UUID to str
    user_id: str
    influencer_id: str  # Changed from UUID to str to support IC Principal IDs
    created_at: datetime
    updated_at: datetime
    metadata: dict[str, object] = Field(default_factory=dict)

    influencer: AIInfluencer | None = None
    message_count: int | None = None
    last_message: LastMessageInfo | None = None
    recent_messages: list[Message] | None = None


class Message(BaseModel):
    """Message entity"""

    model_config = ConfigDict(from_attributes=True)

    id: str  # Changed from UUID to str
    conversation_id: str  # Changed from UUID to str
    role: MessageRole
    content: str | None = None
    message_type: MessageType
    media_urls: list[str] = Field(default_factory=list, description="List of storage keys for image files")
    audio_url: str | None = Field(None, description="Storage key for audio file")
    audio_duration_seconds: int | None = None
    token_count: int | None = None
    created_at: datetime
    metadata: dict[str, object] = Field(default_factory=dict)

File: src/models/entities.rs
use chrono::NaiveDateTime;
use serde::{Deserialize, Serialize};
use strum::{AsRefStr, Display, EnumString};
use utoipa::ToSchema;

// ── Enums ──

#[derive(
    Debug,
    Clone,
    Serialize,
    Deserialize,
    PartialEq,
    sqlx::Type,
    Display,
    EnumString,
    AsRefStr,
    ToSchema,
)]
#[sqlx(rename_all = "lowercase")]
#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
pub enum MessageType {
    #[serde(rename = "text")]
    Text,
    #[serde(rename = "multimodal")]
    Multimodal,
    #[serde(rename = "image")]
    Image,
    #[serde(rename = "audio")]
    Audio,
}

#[derive(
    Debug,
    Clone,
    Serialize,
    Deserialize,
    PartialEq,
    sqlx::Type,
    Display,
    EnumString,
    AsRefStr,
    ToSchema,
)]
#[sqlx(rename_all = "lowercase")]
#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
pub enum MessageRole {
    #[serde(rename = "user")]
    User,
    #[serde(rename = "assistant")]
    Assistant,
}

#[derive(
    Debug, Clone, Serialize, Deserialize, PartialEq, Display, EnumString, AsRefStr, ToSchema,
)]
#[strum(serialize_all = "snake_case")]
pub enum InfluencerStatus {
    #[serde(rename = "active")]
    Active,
    #[serde(rename = "coming_soon")]
    ComingSoon,
    #[serde(rename = "discontinued")]
    Discontinued,
}

// ── Entities ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AIInfluencer {
    pub id: String,
    pub name: String,
    pub display_name: String,
    pub avatar_url: Option<String>,
    pub description: Option<String>,
    pub category: Option<String>,
    pub system_instructions: String,
    pub personality_traits: serde_json::Value,
    pub initial_greeting: Option<String>,
    pub suggested_messages: Vec<String>,
    pub is_active: InfluencerStatus,
    pub is_nsfw: bool,
    pub parent_principal_id: Option<String>,
    pub source: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    pub metadata: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub conversation_count: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_count: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Conversation {
    pub id: String,
    pub user_id: String,
    pub influencer_id: String,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
    pub metadata: serde_json::Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub influencer: Option<AIInfluencer>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_count: Option<i64>,
    pub unread_count: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_message: Option<LastMessageInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recent_messages: Option<Vec<Message>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct LastMessageInfo {
    pub content: Option<String>,
    pub role: MessageRole,
    pub created_at: NaiveDateTime,
    #[schema(default = "delivered")]
    pub status: Option<String>,
    pub is_read: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: String,
    pub conversation_id: String,
    pub role: MessageRole,
    pub content: Option<String>,
    pub message_type: MessageType,
    pub media_urls: Vec<String>,
    pub audio_url: Option<String>,
    pub audio_duration_seconds: Option<i32>,
    pub token_count: Option<i32>,
    pub client_message_id: Option<String>,
    pub created_at: NaiveDateTime,
    pub metadata: serde_json::Value,
    pub status: String,
    pub is_read: bool,
}
